plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java

```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+
+                assertJoinFullyPushedDown(session, leftTableName, rightTableName, "INNER JOIN", "CAST(l.varchar_50 AS VARCHAR(100)) = r.varchar_100", List.of(1));
+                assertJoinFullyPushedDown(session, leftTableName, rightTableName, "INNER JOIN", "l.varchar_50 = CAST(r.varchar_100 AS VARCHAR(50))", List.of(1));
+                assertJoinFullyPushedDown(session, leftTableName, rightTableName, "INNER JOIN", "CAST(l.varchar_50 AS VARCHAR(200)) = CAST(r.varchar_100 AS VARCHAR(200))", List.of(1));
+            }
+        }
+
+        private static List<CastTestCase> supportedCastTypePushdown()
+        {
+            return ImmutableList.of(
+                    new CastTestCase("char(10)", "char(20)"),
+                    new CastTestCase("char(10)", "varchar(20)"),
+                    new CastTestCase("char(10)", "varchar(1001)"),
+                    new CastTestCase("char(10)", "varchar"),
+                    new CastTestCase("char(501)", "char(20)"),
+                    new CastTestCase("char(501)", "varchar(20)"),
+                    new CastTestCase("char(501)", "varchar(1001)"),
+                    new CastTestCase("char(501)", "varchar"),
+                    new CastTestCase("varchar(10)", "char(20)"),
+                    new CastTestCase("varchar(10)", "varchar(20)"),
+                    new CastTestCase("varchar(10)", "varchar(1001)"),
+                    new CastTestCase("varchar(10)", "varchar"),
+                    new CastTestCase("varchar(1001)", "char(20)"),
+                    new CastTestCase("varchar(1001)", "varchar(20)"),
+                    new CastTestCase("varchar(1001)", "varchar(1010)"),
+                    new CastTestCase("varchar(1001)", "varchar"),
+                    new CastTestCase("varchar", "char(20)"),
+                    new CastTestCase("varchar", "varchar(20)"),
+                    new CastTestCase("varchar", "varchar(1001)"));
+        }
+
+        private static List<CastTestCase> unsupportedCastTypePushdown()
+        {
+            return ImmutableList.of(
+                    new CastTestCase("char(10)", "char(501)"),
+                    new CastTestCase("char(501)", "char(520)"),
+                    new CastTestCase("varchar(10)", "char(501)"),
+                    new CastTestCase("varchar(1001)", "char(501)"),
+                    new CastTestCase("varchar", "char(501)"),
+                    new CastTestCase("integer", "tinyint"),
+                    new CastTestCase("integer", "smallint"),
+                    new CastTestCase("integer", "bigint"),
+                    new CastTestCase("integer", "real"),
+                    new CastTestCase("integer", "double"),
+                    new CastTestCase("integer", "decimal(15)"),
+                    new CastTestCase("integer", "decimal(10, 2)"),
+                    new CastTestCase("integer", "decimal(30, 2)"),
+                    new CastTestCase("smallint", "integer"),
```

Comment:
Despite the complexity of the implementation 
https://github.com/trinodb/trino/pull/22728/commits/f4204f3c75cd0510e91e22e1dc8f625eb4229744#r1683242993
I guess these return types can be handled too. Could you please try?